Open Input and Output Stream

DeKaylaPhillips opened the pull request:
<!--
     Before submitting a Pull Request, please ensure you've done the following:
     - 👷‍♀️ Create small PRs. In most cases this will be possible.
     - ✅ Provide tests for your changes.
     - 📝 Use descriptive commit messages.
     - 📗 Update any related documentation and include any relevant screenshots.
-->
# Open Input and Output Stream 

## Related Tickets & Documents
![Screen Shot 2023-07-21 at 1 15 29 PM](https://github.com/DeKaylaPhillips/Echo-Server/assets/91583190/93efb097-6ba5-40f1-bd88-7a1478ebc813)

- Ticket: https://trello.com/c/IyA9R6Ye/39-open-input-and-output-stream

## Type of PR <!-- (check all applicable) -->

- [X] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [X] Documentation Update

## Description
This branch refactors the TCP server to allow a client connected at localhost:3000 to send raw data as a message on the socket back to the server once the connection has been established. Once the client has sent a message back to the server, the client's connection is closed. The documentation has also been updated to reflect the new implementation of the TCP server which now uses the NET Node.js library, and removes references to the previous implementation of the server which used the Express.js framework.

**Source Code**

> <img width="449" alt="Screen Shot 2023-07-21 at 1 12 24 PM" src="https://github.com/DeKaylaPhillips/Echo-Server/assets/91583190/f05a3647-4a40-4da4-b853-db6e56742968">

**Client Terminal**

> <img width="503" alt="Screen Shot 2023-07-21 at 1 19 38 PM" src="https://github.com/DeKaylaPhillips/Echo-Server/assets/91583190/0f9959fe-b0a8-4b33-ae81-fdd3b0507a52">

**Server Terminal**

> <img width="467" alt="Screen Shot 2023-07-21 at 1 18 01 PM" src="https://github.com/DeKaylaPhillips/Echo-Server/assets/91583190/002b35b9-d168-4f4b-8da3-5544093b10ee">

## Added/updated tests?

- [X] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

<img width="513" alt="Screen Shot 2023-07-21 at 1 14 38 PM" src="https://github.com/DeKaylaPhillips/Echo-Server/assets/91583190/fdcbfabf-f0f8-4be1-831a-9f6cdc828fcb">

Commit:
Add readable stream and method to close client connection

Added 'data' event and callback function to log the message sent from a client to the server, and method to close the connection when the client has written to the server. Refactored log statements and added write() method stream data prompt to format the display of the TCP data sent from the server to the client and vice versa.

Commit:
Refactor mock socket class and add tests

Added custom types to utilize in mock socket class and mock instance of net.Server. Refactored the MockSocket class to include the 'on()' emitter function and 'end()' socket function which simulates the behavior resulting from the refactored source code calling these methods. Refactored tests to account for changes in source code, and added new tests to test for the resulting behavior of receiving data from a client and closing a client connection once the data has been received.

Commit:
Update README.md

Naomi-Dennis commented on tests/server/server.spec.ts:
```diff
@@ -1,21 +1,38 @@
   type MaybeVoidFunction = (() => void) | undefined
-  type SocketCallback = ((socket: any) => void) | undefined
+  type MaybeServerCallback = ((socket: any) => void) | undefined
+  type StringOrUint8Arr = string | Uint8Array
+  type WriteFunction = (data: StringOrUint8Arr) => boolean
+  type SocketEventEmitter = (
+    eventName: string,
+    eventListener: (data: Buffer) => void
+  ) => void
+  type EndFunction = (
+    buffer: StringOrUint8Arr,
+    callback: MaybeVoidFunction
+  ) => void
```
🎉 Nice, good job keeping up with the types!

DeKaylaPhillips replied:
Thanks for introducing me to the type aliases, Naomi! Really does help me with simplifying things.

Naomi-Dennis commented on tests/server/server.spec.ts:
```diff
@@ -1,21 +1,38 @@
 jest.mock('net', () => {
   type MaybeNumber = number | undefined
   type MaybeString = string | undefined
   type MaybeVoidFunction = (() => void) | undefined
```
This can be changed in the next PR. There's a built in VoidFunction type. That can be used instead of `(() => void)`

DeKaylaPhillips replied:
Got it! I'll be sure to keep that in mind for the next one. :)

Naomi-Dennis commented on tests/server/server.spec.ts:
```diff
@@ -78,14 +95,35 @@ describe('Server', () => {
     server.openTCPConnection()
-    expect(logMock).toHaveBeenCalledTimes(2)
-    expect(logMock).toHaveBeenCalledWith('Client has connected.')
+    expect(logMock).toHaveBeenCalledTimes(3)
+    expect(logMock).toHaveBeenCalledWith('Client has connected.\n')
```
Hmm, this is starting to feel a bit out of hand. It may be time to refactor a bit since really, we're just expecting 1 log to the server not 3.

DeKaylaPhillips replied:
The tests now only check for 1 console log instead of 3. :)

Commit:
Refactor server & tests

Removed console statements and refactored formatting in source code. Tests have been refactored to use new implementation.
